modules/apps/portal-language/portal-language-extender/src/main/java/com/liferay/portal/language/extender/internal/LanguageClientExtension.java

```diff
@@ -0,0 +1,200 @@
+			StringPool.BLANK);
+
+		if (Validator.isNull(headers.get("Liferay-Client-Extension-Batch")) ||
+			_isAlreadyProcessed(bundle)) {
+
+			return null;
+		}
+
+		Enumeration<URL> enumeration = bundle.findEntries(
+			headers.get("Liferay-Client-Extension-Batch"),
+			"Language_*.properties", true);
+
+		while ((enumeration != null) && enumeration.hasMoreElements()) {
+			URL url = enumeration.nextElement();
+
+			File file = new File(url.getFile());
+
+			String fileName = file.getName();
+
+			try {
+				Company company = _companyLocalService.getCompanyByWebId(
+					PropsUtil.get(PropsKeys.COMPANY_DEFAULT_WEB_ID));
+
+				long companyId = company.getCompanyId();
+
+				String languageId = StringUtil.removeSubstring(
+					fileName, "Language_");
+
+				languageId = StringUtil.removeSubstring(
+					languageId, ".properties");
+
+				URLConnection urlConnection = url.openConnection();
+
+				Reader reader = new InputStreamReader(
+					urlConnection.getInputStream(), StandardCharsets.UTF_8);
+
+				Properties properties = new Properties();
+
+				properties.load(reader);
+
+				User user = _userLocalService.getUserByScreenName(
+					companyId,
+					PropsUtil.get(PropsKeys.DEFAULT_ADMIN_SCREEN_NAME));
+
+				_ploEntryLocalService.importPLOEntries(
+					companyId, languageId, properties, user.getUserId());
+
+				if (_log.isInfoEnabled()) {
+					_log.info(
+						"Processed language file \"" + fileName +
+							"\" successfully");
+				}
+			}
+			catch (PLOEntryImportException.InvalidTranslations
+						ploEntryImportException) {
+
+				for (Throwable throwable :
+						ploEntryImportException.getSuppressed()) {
+
+					_log.error(
+						StringBundler.concat(
+							"Unable to process language file \"", fileName,
+							"\". ", throwable.getMessage(), StringPool.PERIOD));
+				}
+			}
+			catch (IOException | PortalException exception) {
+				_log.error(
+					"Unable to process language file \"" + fileName + "\"",
+					exception);
+			}
+		}
+
+		return bundle;
+	}
+
+	@Override
+	public void modifiedBundle(
+		Bundle bundle, BundleEvent event, Bundle object) {
+	}
+
+	@Override
+	public void removedBundle(Bundle bundle, BundleEvent event, Bundle object) {
+	}
+
+	@Activate
+	protected void activate(
+		BundleContext bundleContext, Map<String, Object> properties) {
+
+		_bundleTracker = new BundleTracker<>(
+			bundleContext, Bundle.ACTIVE, this);
+
+		_bundleTracker.open();
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		_bundleTracker.close();
+	}
+
+	private boolean _isAlreadyProcessed(Bundle bundle) {
+		File file = bundle.getDataFile(".liferay-client-extension-language");
+
+		String lastModifiedString = String.valueOf(bundle.getLastModified());
+
+		try {
+			if ((file != null) && file.exists() &&
+				Objects.equals(FileUtil.read(file), lastModifiedString)) {
+
+				return true;
+			}
+
+			if (!file.exists()) {
+				file.createNewFile();
+			}
+
+			FileUtil.write(file, lastModifiedString, true);
```

Comment:
It's a good practice to avoid mutate/create state in accessor method, i.e `getXXX()`,` isXXXX()`

Reply:
Will change it in the next PR, I don't want to run CI again (it will take hours) :sweat_smile: